system-addon/content-src/components/TopSites/TopSites.jsx

```diff
@@ -71,22 +71,28 @@ class TopSites extends React.PureComponent {
   render() {
     const props = this.props;
     const topSites = this._getTopSites();
-
-    const placeholderCount = props.TopSitesCount - topSites.length;
     const infoOption = {
       header: {id: "settings_pane_topsites_header"},
       body: {id: "settings_pane_topsites_body"}
     };
+    const topSitesUI = [];
+    for (let i = 0, l = props.TopSitesCount; i < l; i++) {
+      const link = topSites[i];
+      if (link) {
+        topSitesUI.push(<TopSite
+          key={link.guid || link.url}
+          dispatch={props.dispatch}
+          link={link}
+          index={i}
+          intl={props.intl} />);
+      } else {
+        topSitesUI.push(<TopSitePlaceholder key={i} />);
```

Comment:
nit: could do `topSitesUI.push(!link ? <TopSitePlaceholder… > : …)`

A quick test also verifies that this does what it should do, but yeah some refactoring would be much nicer